src/order/has_involution.lean

```diff
@@ -0,0 +1,200 @@
+/-
+Copyright (c) 2022 Peter Nelson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Peter Nelson
+-/
+import order.antichain
+
+/-!
+# Orders with involution
+
+This file concerns orders that admit an order-reversing involution. In the case of a lattice,
+these are sometimes referred to as 'i-lattices' or 'lattices with involution'. Such an involution
+is more general than a `boolean_algebra` complement, but retains many of its properties. Other than
+a boolean algebra, an example is the subspace lattice of the vector space `𝕂ⁿ` for `𝕂` of nonzero
+characteristic, where for each subspace `W` we have `invo W = {x ∈ V | ∀ w ∈ W, wᵀx = 0}`; this is
+not a complement in the stronger sense because `invo W` can intersect `W`.
+
+## Main declarations
+
+* `has_involution`: typeclass applying to types with a `preorder` that admit an antitone involution.
+
+* `ⁱ` : postfix notation for the function `invo : α → α` given a type `α` with `[has_involution α]`
+
+## TODO
+
+Provide instances other than the one from `boolean_algebra`.
+-/
+
+universe u
+
+/-- An antitone involution on a preorder -/
+class has_involution (α : Type u) [preorder α]  :=
+(invo : α → α)
+(invo_antitone' : ∀ (x y : α), x ≤ y → invo y ≤ invo x)
+(invo_involutive' : function.involutive invo)
+
+open has_involution
+
+variables {α : Type u}
+
+postfix `ⁱ`:(max+1) := invo
+
+section preorder
+
+variables [preorder α] [has_involution α] {x y : α}
+
+@[simp] lemma invo_invo (x : α) : xⁱⁱ = x :=  invo_involutive' x
+
+lemma invo_eq_iff_invo_eq : xⁱ = y ↔ yⁱ = x :=
+by {rw [eq_comm], exact invo_involutive'.eq_iff.symm}
+
+lemma eq_invo_iff_eq_invo : x = yⁱ ↔ y = xⁱ :=
+by rw [← invo_invo x, invo_eq_iff_invo_eq, invo_invo, invo_invo]
+
+lemma invo_le_invo (h : x ≤ y) : yⁱ ≤ xⁱ := invo_antitone' _ _ h
+
+lemma le_of_invo_le (h : xⁱ ≤ yⁱ) : y ≤ x :=
+by {rw [←invo_invo x, ←invo_invo y], exact invo_le_invo h,}
+
+lemma invo_le_invo_iff_le : xⁱ ≤ yⁱ ↔ y ≤ x := ⟨le_of_invo_le, invo_le_invo⟩
+
+lemma le_invo_iff_le_invo : x ≤ yⁱ ↔ y ≤ xⁱ := by rw [←invo_le_invo_iff_le, invo_invo]
+
+lemma invo_le_iff_invo_le : xⁱ ≤ y ↔ yⁱ ≤ x := by rw [←invo_le_invo_iff_le, invo_invo]
+
+lemma invo_inj (h : xⁱ = yⁱ) : x = y := invo_involutive'.injective h
```

Comment:
I don't think there's a need to have this lemma, when `invo_injective` does the same thing.